### DISCUSSION TITLE
Implementing many-to-many relationships with the UI

### DISCUSSION THREAD

**Discussion by gaurangrshah** [supabase/supabase#710] [Questions] [answered]
I must be missing something, but I'm currently experimenting with supabase, and I can't seem to figure out how to create many-to-many or one-to-many relationships using the UI. Are there any blog posts, that discuss this? I am more than sure that I'm missing something. I've read through the docs pretty much cover -to-cover and can't seem to find anything other than this: https://supabase.io/docs/guides/database#relationships

Any help/suggestions/guidance would be greatly appreciated! Thanks in advance for the help!

**Comment by kiwicopple** [comment, marked as answer]
Hey @gaurangrshah, a many-to-many relationship is created by essentially creating 2 tables, then a final "join" table to connect these together. 


https://user-images.githubusercontent.com/10214025/107771512-d1990600-6d75-11eb-9bb1-9ceb4a5c8fd9.mov

**Reply by manuelbiermann** [reply]
Thank you for this video @kiwicopple. Is there something similar describing on how to query supabase for all users which are part of a Set of teams?

**Reply by kiwicopple** [reply]
If you followed this tutorial @manuelbiermann , it should be as simple as this:

```js
const { data: teams, error } = supabase
  .from('teams')
  .select(`
    id, name,
    users ( id, email )
  `)
```

Or in reverse:

```js
const { data: users, error } = supabase
  .from('users')
  .select(`
    id, email,
    teams ( id, name )
  `)
```

**Reply by manuelbiermann** [reply]
I think my question was not clearly formulated. What I am trying to achieve is to query for example for all users which are part of teams 3, 4, and 5.

**Reply by pranavbadami** [reply]
Hey @manuelbiermann , did you ever find a solution for querying in this way? Running into the same issue and I'd ideally like to perform the query using the Supabase JS library as opposed to an `rpc`. Thanks!

**Reply by pranavbadami** [reply]
The best I've been able to do so far is the following: 
```
const {data, error} = await supaClient.from("users").select("name, tags (id, name)").in("tags.id", [93, 92])
const test = data.filter(row => row.teams.length > 0)
```

which returns something like this:
```
[{
  name: 'Chris',
  tags: [ { id: 93, name: 'Team A' }, { id: 92, name: 'Team B' } ]
},
{
 name: 'John',
  tags: [ { id: 93, name: 'Team A' }]
},
...]
```
I'd ideally like to combine this into one query, but right now I have to query and filter on the client side.

**Reply by waspeer** [reply]
How in this example (users, teams, members) would you do an insert/upsert with the supabase javascript client? Maybe I'm missing something, but I'm unable to get it to work.

**Reply by wilfredinni** [reply]
> How in this example (users, teams, members) would you do an insert/upsert with the supabase javascript client? Maybe I'm missing something, but I'm unable to get it to work.

@waspeer did you find out?

**Reply by waspeer** [reply]
I did not. I'm assuming it's not possible in one api call, so I'm splitting it up into multiple calls. Eg:

```js
const usersResponse = await client.from('users').upsert([{ email: '...' }, { email: '...' }]);
const teamResponse = await client.from('teams').insert({ name: 'TeamName' });
const teamId = teamResponse.data[0].id;

await client.from('members').insert(
  usersResponse.data.map(({id}) => ({ user_id: id, team_id: teamId })),
);
```

_(Disclaimer, I quickly adapted this code from my project and removed stuff like error handling for simplicity, so it's untested)_

It would be great if it was possible in one api call though! Seems like this is possible for 1-m relationships. If someone knows a way, please advice.

**Reply by aditodkar** [reply]
Is it possible to use upsert with many to many relationship? In such scenario should I use upsert().match() and inside match query will be those attributes on basis of which I want to upsert. Is this correct way? Any other suggestions?

**Reply by tomaaron** [reply]
> If you followed this tutorial @manuelbiermann , it should be as simple as this:
> 
> ```js
> const { data: teams, error } = supabase
>   .from('teams')
>   .select(`
>     id, name,
>     users ( id, email )
>   `)
> ```
> 
> Or in reverse:
> 
> ```js
> const { data: users, error } = supabase
>   .from('users')
>   .select(`
>     id, email,
>     teams ( id, name )
>   `)
> ```

I'm having issue with this and I'm getting the following error:
```
Could not find a relationship between 'teams' and 'users' in the schema cache
```
Does somebody has an idea?

**Reply by derek-rein** [reply]
Can this work in a one-to-many relationship without a join table?

**Reply by awaaisali** [reply]
This query isn't working for me. Can you please give any reference related to that?

**Reply by GaryAustin1** [reply]
> This query isn't working for me. Can you please give any reference related to that?

@awaaisali if you are using a join table with two fk's pointing to two other tables (it is unclear with so many comments here what your table structure or issue is) then you need to declare both key columns in the join table to be part of primary key.  This was a recent change to PostgREST.  In the UI checking them both primary does this.

**Reply by awaaisali** [reply]
@GaryAustin1  The problem is solved by using this query.
const { data, error } = await supabase
  .from('users')
  .select(' *, members!inner ( * ) ')

**Reply by hansaliyad1** [reply]
I am having the same issue but @awaaisali solution doesn't work for me. I am not sure what am I missing.

**Reply by GaryAustin1** [reply]
@hansaliyad1 This is a pretty old issue, with alot of different comments added to be joining.
There could be several different things you are doing wrong as you don't really say what your issue is and set up is.
Did you make sure both foreign keys are part of the primary key?

**Reply by hansaliyad1** [reply]
Hey @GaryAustin1,

This is my first time posting to this community. I have been working on Supabase with Flutter for the past 2 months, so I am fairly new. I am encountering an error when trying to run the query below. The error I'm getting is: `'Could not find a relationship between 'profiles' and 'workflows' in the schema cache.'`

```
final data = await _supabase 
  .from('profiles') 
  .select(', workflows()');
```
I have three tables that set up a many-to-many relationship:

1. 'profiles'
2. 'workflows'
3. 'rel_profiles_workflows'

I am pretty sure I am missing something here, but I'm not sure what it is. I have attached images below to show you how my tables and relationships are set up.

Profiles table:
![image](https://github.com/supabase/supabase/assets/9779007/74e5ce61-356e-4fcf-b09b-c7c3772d682b)

Workflow table:
![image](https://github.com/supabase/supabase/assets/9779007/72bde7ce-73d6-4eda-be6c-06d0728ea717)

rel_profiles_workflows table:
![image](https://github.com/supabase/supabase/assets/9779007/ef9b4759-2c35-43fb-ae05-53349816e897)

Relationship between profile and rel_profiles_workflows table:
![image](https://github.com/supabase/supabase/assets/9779007/badf9ab7-6c5f-4137-86b5-189358a05ff0)

Relationship between workflows and rel_profiles_workflows table:
![image](https://github.com/supabase/supabase/assets/9779007/a8395104-7e98-48bb-9e1c-f53a1d71c3cc)

I would appreciate any advice or assistance in this matter. Thank you!

**Reply by GaryAustin1** [reply]
Both of your foreign keys in rel_profiles_workflows need to be marked as part of the primary key (there can be multiple columns in the primary key).  This is  a requirement for the API to look at them in a join table.

**Reply by hansaliyad1** [reply]
@GaryAustin1 You are the Savior!! That was it. I did had to re-create the table to mark the foreign keys as part of the primary key. I was getting `"rel_profiles_workflows_pkey" of relation "rel_profiles_workflows" does not exist` error when I was trying to mark the foreign keys as part of the primary key.

**Reply by visualjerk** [reply]
I ran into the same issue with an old relation table, that could not be updated. The migration was pretty straightforward:

- In "Table Editor" select the old relationship table > select all rows (make sure to click "Select all XXX rows") > Export to CSV
- Create the new relationship table with the same row names as the old one and foreign keys marked as part of the primary key
- Select the new table > Insert > Import data from CSV
- If needed: Copy the RLS policies from the old table to the new one

Hope it helps anyone experiencing this issue.

**Comment by gaurangrshah** [comment]
Worked like a charm, thanks...I could swear I tried this, but I def. must've done something wrong in the process. Either way thanks a ton!

**Comment by borngraced** [comment]
@kiwicopple sorry, but what If I want to insert to many-to-many relation using supabase api? currently, I'm able to do that effortlessly with prisma api.


code sample using Prisma

```js
const publish = await prisma.job.create({
      data: {
        title,
        uri,
        body,
        published,
        lead,
        Profile: { connect: { id: userId } },
        expired: false,
        company: { connect: { id: company } },
        contract: { connect: { id: contract } },
        experience: { connect: { id: experience } },
        location: { connect: { id: location } },
        categories: { connect: categories },
        created_at: new Date(),
        updated_at: new Date(),
      },
    });
```

using supabase api

```js
const user: any = supabase.auth.user();
      const newJobDetails: any = {
        title: dataJob.title,
        uri: dataJob.body,
        lead: dataJob.lead,
        body: dataJob.body,
        contract: dataJob.contractId,
        experience: dataJob.experienceId,
        location: dataJob.location,
        company: dataJob.companyId,
        published: dataJob.published,
        userId: user?.id,
        categories: [{"id":2}, {"id":4}],
        created_at: new Date(),
        updated_at: new Date(),
      };

      const { data, error } = await supabase
        .from("job")
        .insert([newJobDetails]);
```

**Reply by wilfredinni** [reply]
@Samuelonoja did you find out?

**Reply by trinhdh** [reply]
@Samuelonoja did you find out ?

**Reply by borngraced** [reply]
@wilfredinni  @trinhdh 

No, not yet.. the document isn't clear or didn't give an highlight on how to do that. which can be very frustrating.

but I was able to implement updating && writing to the DB including many-to-many relations using Prisma ORM

**Reply by soedirgo** [reply]
I may be missing something, but inserting to many-to-many relationship is just inserting into the join table right?

**Reply by borngraced** [reply]
Yes but programmatically.

**Reply by trinhdh** [reply]
@Samuelonoja Yeah. Thanks. Hope supabase will have it in the future release. What is your experience using Prisma ORM ? i never tried that before.

**Reply by borngraced** [reply]
Oh,  Prisma ORM I'll consider a great tool....I'd def recommend it to anyone

**Reply by viper-42** [reply]
This is the first page that came up when searching for a way to connect many to many relationships through the supabase api like in prisma so I thought I would share the solution I found. In order to do it with 1 query from the client we need to create a postgres function as a remote procedure call. Here is what I came up with for a todolist to create a task with related tags:
```sql
CREATE OR REPLACE FUNCTION add_task_with_tags (input_owner_id uuid, input_title text, input_description text DEFAULT NULL, input_priority text DEFAULT NULL, input_due_by timestamp DEFAULT NULL, input_tag_ids int[] DEFAULT NULL)
    RETURNS SETOF tasks
    AS $$
DECLARE
    new_task_id int;
    input_tag_id int;
BEGIN
    INSERT INTO tasks (owner_id, title, description, priority, due_by)
        VALUES (input_owner_id, input_title, input_description, input_priority, input_due_by)
    RETURNING
        id INTO new_task_id;
    IF input_tag_ids IS NOT NULL THEN
        foreach input_tag_id IN ARRAY input_tag_ids LOOP
            INSERT INTO tags_to_tasks (task_id, tag_id)
                VALUES (new_task_id, input_tag_id);
        END LOOP;
    END IF;
    RETURN query
    SELECT
        *
    FROM
        tasks
    WHERE
        tasks.id = new_task_id;
END;
$$
LANGUAGE plpgsql;
```
